ang/civiawards/shared/directives/contacts-group-dropdown.directive.js

```diff
@@ -0,0 +1,193 @@
+        });
+
+        elem.on('select2-selecting', selectEventHandler);
+
+        $(elem).on('select2-removing', removeEventHandler);
+      }
+
+      /**
+       * @param {Array} groupsData groups data
+       */
+      function prepareGroupsDataForSelect2 (groupsData) {
+        _.chain(groupsData)
+          .filter(function (group) {
+            return group.extra.is_hidden === '0' && group.extra.is_active === '1';
+          })
+          .each(function (group) {
+            scope.groups.results[0].children.push({
+              id: 'i' + group.id,
+              value: group.id,
+              mode: 'include',
+              text: group.label
+            });
+            scope.groups.results[1].children.push({
+              id: 'e' + group.id,
+              value: group.id,
+              mode: 'exclude',
+              text: group.label
+            });
+          }).value();
+      }
+
+      /**
+       * Fetch Groups data using API
+       *
+       * @returns {Promise} promise
+       */
+      function fetchGroupsData () {
+        return crmApi('Group', 'getlist', {
+          options: { limit: 0 },
+          extra: ['is_hidden', 'is_active']
+        }).then(function (groupsData) {
+          return groupsData.values;
+        });
+      }
+
+      /**
+       * Select event handler for dropdown
+       *
+       * @param {object} e event object
+       */
+      function selectEventHandler (e) {
+        if (e.object.mode === 'exclude') {
+          ngModel.$modelValue.exclude.push(e.object.value);
+          arrayRemove(ngModel.$modelValue.include, e.object.value);
+        } else {
+          ngModel.$modelValue.include.push(e.object.value);
+          arrayRemove(ngModel.$modelValue.exclude, e.object.value);
+        }
+
+        scope.$apply();
+        $(elem).select2('close');
+        e.preventDefault();
+      }
+
+      /**
+       * Remove event handler for dropdown
+       *
+       * @param {object} e event object
+       */
+      function removeEventHandler (e) {
+        arrayRemove(ngModel.$modelValue[e.choice.mode], e.choice.value);
+
+        scope.$parent.$apply();
+
+        e.preventDefault();
+      }
+
+      /**
+       * Show selected values in dropddown
+       */
+      function refreshUI () {
+        scope.recips = ngModel.$modelValue;
+        if (ngModel.$modelValue) {
+          $(elem).select2('val', convertGroupsToValues(ngModel.$modelValue));
+        }
+      }
+
+      /**
+       * Convert selected values to the format suitable for Select2
+       *
+       * @param {object} values selected values
+       * @returns {Array} selected values in format suitable for Select2
+       */
+      function convertGroupsToValues (values) {
+        var r = [];
+
+        _.each(values.include, function (v) {
+          r.push('i' + v);
+        });
+
+        _.each(values.exclude, function (v) {
+          r.push('e' + v);
+        });
+
+        return r;
+      }
```

Comment:
Several issues here:
- Can we remove these single variables and more descriptive ones?
- Why `'i' + v` ? can it be `'include_' + contact_id` etc ?

I think that can be refactored like this:

```js
function convertGroupsToValues (values) {
  return _.concat([],
    _.map(values.include, function (contact_id) { return 'include_' + contact_id; }),
    _.map(values. exclude, function (contact_id) { return 'exclude_' + contact_id; })
  );
}
```